chain/src/epoch.rs

```diff
@@ -1,62 +1,39 @@
-use tendermint::block;
+use penumbra_proto::{core::chain::v1alpha1 as pb_chain, DomainType};
+use serde::{Deserialize, Serialize};
 
 /// Penumbra groups blocks into epochs and restricts validator changes to epoch boundaries.
-///
-/// The epoch duration is a chain parameter and cannot be changed after initialization.
-#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Copy, Clone)]
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Serialize, Deserialize)]
+#[serde(try_from = "pb_chain::Epoch", into = "pb_chain::Epoch")]
 pub struct Epoch {
     pub index: u64,
-    pub duration: u64,
+    pub start_height: u64,
 }
 
-impl Epoch {
-    /// Instantiates a new `Epoch` from a given block height and epoch duration.
-    pub fn from_height(height: u64, epoch_duration: u64) -> Epoch {
-        Epoch {
-            index: height / epoch_duration,
-            duration: epoch_duration,
-        }
-    }
-
-    /// Indicates the starting block height for this epoch (inclusive)
-    pub fn start_height(&self) -> block::Height {
-        block::Height::try_from(self.index * self.duration).expect("able to parse block height")
-    }
-
-    /// Indicates the ending block height for this epoch (inclusive)
-    pub fn end_height(&self) -> block::Height {
-        block::Height::try_from((self.index + 1) * self.duration - 1)
-            .expect("able to parse block height")
-    }
-
-    pub fn is_epoch_end(&self, height: u64) -> bool {
-        self.end_height().value() == height
-    }
+impl DomainType for Epoch {
+    type Proto = pb_chain::Epoch;
+}
 
-    /// Returns the epoch following this one.
-    pub fn next(&self) -> Self {
+impl From<pb_chain::Epoch> for Epoch {
+    fn from(msg: pb_chain::Epoch) -> Self {
         Epoch {
-            index: self.index + 1,
-            duration: self.duration,
+            index: msg.index,
+            start_height: msg.start_height,
         }
     }
+}
 
-    /// Returns the epoch preceding this one.
-    pub fn prev(&self) -> Self {
-        Epoch {
-            index: self.index - 1,
-            duration: self.duration,
+impl From<Epoch> for pb_chain::Epoch {
+    fn from(epoch: Epoch) -> Self {
+        pb_chain::Epoch {
+            index: epoch.index,
+            start_height: epoch.start_height,
         }
     }
 }
 
-impl std::ops::Add<u64> for Epoch {
-    type Output = Epoch;
-
-    fn add(self, other: u64) -> Self::Output {
-        Epoch {
-            index: self.index + other,
-            duration: self.duration,
-        }
+impl Epoch {
+    // NOTE: this should be changed when we move to time-based mechanisms
+    pub fn is_epoch_end(&self, current_height: u64, epoch_duration: u64) -> bool {
```

Comment:
Even if we're planning to change it later, maybe this should be `is_scheduled_epoch_end`, to avoid confusion? Or we could move the logic into a private method on the `App`, so that we can't have any other part of the code predict when epochs will end (and thereby become surprised if they end early).